### PR TITLE
Add test to check connection to the apiserver

### PR DESCRIPTION
**Pull request by rajansandeep**
Adds test to ensure coredns does not crash when the apiserver is restarted. 
This test validates https://github.com/coredns/deployment/pull/138 and catch issues like https://github.com/coredns/coredns/issues/2464

### CODE REVIEW

File: test/k8sdeployment/connection_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		t.Fatalf("deployment script failed: %s\nerr: %s", string(cmdout), err)
+	}
+
+	// Restart the Kubernetes APIserver and wait for it to come back up.
+	dockerCmd, err := exec.Command("sh", "-c", "docker restart $(docker ps --no-trunc | grep 'kube-apiserver' | awk '{ print $1; }') > /dev/null").CombinedOutput()
+	if err != nil {
+		t.Fatalf("docker container restart failed: %s\nerr: %s", string(dockerCmd), err)
+	}
+	time.Sleep(15 * time.Second)
```
Comment (chrisohaver) [inline test/k8sdeployment/connection_test.go]: OK - if total duration of the deployment test becomes an issue we should revisit this.  The crash should be immediate, so we should be able to detect the occurrence in a lot less than 15 seconds.  Can this be lowered at all without flaking?

File: test/k8sdeployment/connection_test.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+	// Get the restart count of the CoreDNS pods.
+	restartCount, err := kubernetes.Kubectl("-n kube-system get pods -l k8s-app=kube-dns -ojsonpath='{.items[*].status.containerStatuses[0].restartCount}'")
+	if err != nil {
+		t.Fatalf("error fetching CoreDNS pod restart count: %s", err)
+	}
+
+	// If CoreDNS crashes due to KubeAPIServer restart, Kubernetes will restart the CoreDNS containers.
+	if !strings.Contains(restartCount, "0") {
+		t.Fatalf("failed as CoreDNS crashed due to KubeAPIServer restart.")
```
Comment (chrisohaver) [inline test/k8sdeployment/connection_test.go]: We know that coredns restarted... we don't know necessarily that it was caused by the API restart.  CoreDNS could have something else wrong, like an invalid config or other bug that causes it to crash repeatedly.

To be more sure, you could add another sleep/test _before_ the api restart to verify the restarts are zero.

File: test/k8sdeployment/connection_test.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+func TestConnectionAfterAPIRestart(t *testing.T) {
+	// Apply manifests via coredns/deployment deployment script ...
+	path := os.Getenv("DEPLOYMENTPATH")
+	cmd := exec.Command("sh", "-c", "./deploy.sh -s -i 10.96.0.10 -r 10.96.0.0/8 -r 172.17.0.0/16 | kubectl apply -f -")
+	cmd.Dir = path + "/kubernetes"
+	cmdout, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("deployment script failed: %s\nerr: %s", string(cmdout), err)
+	}
+
```
Comment (chrisohaver) [inline test/k8sdeployment/connection_test.go]: Probably should wait until the coredns pods come up at this point before doing the API test.

File: test/k8sdeployment/connection_test.go (hunk 1 above)
Comment (rajansandeep) [reply test/k8sdeployment/connection_test.go]: There seems to be a delay from when the apiserver is restarted to getting the output for the number of restarts. I tried the sleep to 10secs, but the test flaked and passed. 15 secs is a lot but is reqd for the test to not flake.

File: test/k8sdeployment/connection_test.go (hunk 2 above)
Comment (rajansandeep) [reply test/k8sdeployment/connection_test.go]: Ahh yes, I'll add that.

**Comment by chrisohaver** [comment]
> Probably should wait until the coredns pods come up at this point before doing the API test.

This may be why you cannot drop the 15 second sleep lower.
The prior test doesn't wait for coredns to come up.  It just looks at the configmap object.
So, docker may still be downloading the coredns image, coredns pods not ready.

**Comment by chrisohaver** [comment]
> This may be why you cannot drop the 15 second sleep lower.
The prior test doesn't wait for coredns to come up. It just looks at the configmap object.
So, docker may still be downloading the coredns image, coredns pods not ready.

Actually, thats not right. With minikube in the vmdriver=none mode (uses local docker), the images would be retained from run to run, so they would not be re-downloading...

File: test/k8sdeployment/connection_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	// Restart the Kubernetes APIServer.
+	dockerCmd, err := exec.Command("sh", "-c", "docker restart $(docker ps --no-trunc | grep 'kube-apiserver' | awk '{ print $1; }') > /dev/null").CombinedOutput()
+	if err != nil {
+		t.Fatalf("docker container restart failed: %s\nerr: %s", string(dockerCmd), err)
+	}
+	// Verify that the CoreDNS pods are up and ready after the restart.
+	maxWait = 120
+	if kubernetes.WaitNReady(maxWait, 2) != nil {
```
Comment (rajansandeep) [inline test/k8sdeployment/connection_test.go]: I removed the 15 sec sleep and added to check if coredns is ready before checking the restart count.
